pkg/capacity/capacity.go

```diff
@@ -0,0 +1,627 @@
+	// We lock only while accessing c.capacities, but not during
+	// the potentially long-running operations. That is okay
+	// because there is only a single worker per item. In the
+	// unlikely case that the desired state of the item changes
+	// while we work on it, we will add or update an obsolete
+	// object which we then don't store and instead queue for
+	// removal.
+	c.capacitiesLock.Lock()
+	capacity, found := c.capacities[item]
+	c.capacitiesLock.Unlock()
+
+	klog.V(5).Infof("Capacity Controller: refreshing %+v", item)
+	if !found {
+		// The item was removed in the meantime. This can happen when the storage class
+		// or the topology segment are gone.
+		klog.V(5).Infof("Capacity Controller: %v became obsolete", item)
+		return nil
+	}
+
+	sc, err := c.scInformer.Lister().Get(item.storageClassName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			// Another indication that the value is no
+			// longer needed.
+			return nil
+		}
+		return fmt.Errorf("retrieve storage class for %+v: %v", item, err)
+	}
+
+	req := &csi.GetCapacityRequest{
+		Parameters: sc.Parameters,
+		// The assumption is that the capacity is independent of the
+		// capabilities. The standard makes it mandatory to pass something,
+		// therefore we pick something rather arbitrarily.
+		VolumeCapabilities: []*csi.VolumeCapability{
+			{
+				AccessType: &csi.VolumeCapability_Mount{},
+				AccessMode: &csi.VolumeCapability_AccessMode{
+					Mode: csi.VolumeCapability_AccessMode_UNKNOWN,
+				},
+			},
+		},
+	}
+	if item.segment != nil {
+		req.AccessibleTopology = &csi.Topology{
+			Segments: item.segment.GetLabelMap(),
+		}
+	}
+	resp, err := c.csiController.GetCapacity(ctx, req)
+	if err != nil {
+		return fmt.Errorf("CSI GetCapacity for %+v: %v", item, err)
+	}
+
+	quantity := resource.NewQuantity(resp.AvailableCapacity, resource.BinarySI)
+	if capacity == nil {
+		// Create new object.
+		capacity = &storagev1alpha1.CSIStorageCapacity{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName:    "csisc-",
+				OwnerReferences: []metav1.OwnerReference{c.owner},
+			},
+			StorageClassName: item.storageClassName,
+			NodeTopology:     item.segment.GetLabelSelector(),
+			Capacity:         quantity,
+		}
+		var err error
+		klog.V(5).Infof("Capacity Controller: creating new object for %+v, new capacity %v", item, quantity)
+		capacity, err = c.client.StorageV1alpha1().CSIStorageCapacities(c.ownerNamespace).Create(ctx, capacity, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("create CSIStorageCapacity for %+v: %v", item, err)
+		}
+		klog.V(5).Infof("Capacity Controller: created %s for %+v with capacity %v", capacity.Name, item, quantity)
+	} else if capacity.Capacity.Value() == quantity.Value() {
+		klog.V(5).Infof("Capacity Controller: no need to update %s for %+v, same capacity %v", capacity.Name, item, quantity)
+		return nil
+	} else {
+		// Update existing object.
+		capacity.Capacity = quantity
```

Comment:
I  think this capacity object can point to the object that's stored in the informer, since it's populated from the event handlers. We should not modify objects stored in the informer cache. Instead we need to make a copy of object before doing an update.

Reply:
Okay, done.